webapp-django/crashstats/signature/static/signature/js/signature_tab_graph.js

```diff
@@ -0,0 +1,203 @@
+ * Controlled by a select.
+ *
+ * @extends {SignatureReport.Tab}
+ * @inheritdoc
+ */
+SignatureReport.GraphTab = function(tabName) {
+
+    var config = {
+        'panels': true,
+        'defaultOptions': ['nightly'],
+        'dataDisplayType': 'graph',
+        'pagination': false
+    };
+
+	SignatureReport.Tab.call(this, tabName, config);
+};
+
+SignatureReport.GraphTab.prototype = SignatureReport.inherit(SignatureReport.Tab.prototype);
+
+// Extends loadControls to load a select with options for each channel.
+SignatureReport.GraphTab.prototype.loadControls = function() {
+
+    // For accessing this inside functions.
+    var that = this;
+
+    // Make the select element and one empty option (for select2).
+    var selectElement = $('<select>', {'class': 'channels-list'});
+    selectElement.append($('<option>'));
+
+    // Make and append and option for each channel.
+    $.each(window.CHANNELS, function (i, channel) {
+        selectElement.append($('<option>', {
+            'value': channel,
+            'text': channel
+        }));
+    });
+
+    // Append the controls.
+    this.$controlsElement.append(selectElement, $('<hr>'));
+
+    // Apply select2 to select element.
+    selectElement.select2({
+        'placeholder': 'Channel...',
+        'allowClear': true
+    });
+
+    // Draw a new graph whenever a new channel is chosen.
+    selectElement.on('change', function (e) {
+        selectElement.select2('val', '');
+        that.loadPanel(e.val);
+    });
+
+};
+
+// Extends getParamsForUrl to handle two particular situations:
+// 1) when multiple products are defined
+// 2) when no product is defined
+SignatureReport.GraphTab.prototype.getParamsForUrl = function () {
+
+    // Get the params as usual.
+    var params = SignatureReport.getParamsWithSignature();
+
+    // If a product has been defined...
+    if (params.product) {
+
+        // ...There may be multiple products, so pick the first one.
+        params.product = params.product[0];
+
+    // If no product has been defined, the user must choose one.
+    } else {
+
+        // Show the search form.
+        var searchFormLink = $('.toggle-filters');
+        searchFormLink.removeClass('show');
+        searchFormLink.html('Hide');
+        SignatureReport.form.show();
+
+        // Add a new fieldset for choosing a product.
+        SignatureReport.form.dynamicForm('newLine', {
+            field: 'product',
+            operator: 'has terms',
+            value: ''
+        });
+
+        // Focus the value input of the new fieldset.
+        $('fieldset:last-child .value').select2('open');
+
+        // Prompt the user to choose a product.
+        this.$contentElement.empty().text('Please select a product.');
+
+    }
+
+    return params;
+
+};
+
+// Extends onAjaxSuccess to process the data and draw a graph.
+SignatureReport.GraphTab.prototype.onAjaxSuccess = function (contentElement, data) {
+
+    // Processes the data.
+    function formatData(data) {
```

Comment:
Why not put those 2 functions in the class as methods and avoid recreating them every time the tab is displayed? It would also be easier to read I think.